cocotb/__init__.py

```diff
@@ -56,36 +56,42 @@
 
 # GPI logging instance
 if "COCOTB_SIM" in os.environ:
-    import simulator
-    logging.basicConfig()
-    logging.setLoggerClass(SimBaseLog)
-    log = SimLog('cocotb')
-    level = os.getenv("COCOTB_LOG_LEVEL", "INFO")
-    try:
-        _default_log = getattr(logging, level)
-    except AttributeError as e:
-        log.error("Unable to set loging level to %s" % level)
-        _default_log = logging.INFO
-    log.setLevel(_default_log)
-    loggpi = SimLog('cocotb.gpi')
-    # Notify GPI of log level
-    simulator.log_level(_default_log)
+
+    def _reopen_stream_with_buffering(stream_name):
+        try:
+            if not getattr(sys, stream_name).isatty():
+                setattr(sys, stream_name, os.fdopen(getattr(sys, stream_name).fileno(), 'w', 1))
+                return True
+            return False
+        except Exception as e:
+            return e
 
     # If stdout/stderr are not TTYs, Python may not have opened them with line
     # buffering. In that case, try to reopen them with line buffering
     # explicitly enabled. This ensures that prints such as stack traces always
     # appear. Continue silently if this fails.
-    try:
-        if not sys.stdout.isatty():
-            sys.stdout = os.fdopen(sys.stdout.fileno(), 'w', 1)
-            log.debug("Reopened stdout with line buffering")
-        if not sys.stderr.isatty():
-            sys.stderr = os.fdopen(sys.stderr.fileno(), 'w', 1)
-            log.debug("Reopened stderr with line buffering")
-    except Exception as e:
-        log.warning("Failed to ensure that stdout/stderr are line buffered: %s", e)
+    _stdout_ok = _reopen_stream_with_buffering('stdout')
+    _stderr_ok = _reopen_stream_with_buffering('stderr')
+
+    # don't set the logging up until we've attempted to fix the standard IO,
+    # otherwise it will end up connected to the unfixed IO.
+    cocotb.log.default_config()
+    log = logging.getLogger(__name__)
+
+    # we can't log these things until the logging is set up!
+    if _stderr_ok is True:
+        log.debug("Reopened stderr with line buffering")
```

Comment:
These log messages are nice, but don't convey much useful information. (What are you doing with that info?) It would simplify the code not to log them.

Reply:
They're marked `debug` because they're probably not very useful :). It seems like for someone debugging why their exception tracebacks are being swallowed, this might provide a little bit of useful info.